src/qs_vxc_atom.F

```diff
@@ -307,24 +305,20 @@ SUBROUTINE calculate_vxc_atom(qs_env, energy_only, exc1, gradient_atom_set, &
                   CALL get_rho_atom(rho_atom=rho_atom, rho_rad_h=r_h, rho_rad_s=r_s)
                END IF
                IF (tau_f) THEN
-                  NULLIFY (trho_h, trho_s)
-                  CALL get_rho_atom(rho_atom=rho_atom, trho_rad_h=trho_h, trho_rad_s=trho_s)
+                  !compute tau on the grid all at once
+                  CALL calc_tau_atom(tau_h, tau_s, rho_atom, my_kind_set(ikind), nspins)
                END IF
 
                DO ir = 1, nr
                   CALL calc_rho_angular(grid_atom, harmonics, nspins, gradient_f, &
                                         ir, r_h, r_s, rho_h, rho_s, dr_h, dr_s, &
                                         r_h_d, r_s_d, drho_h, drho_s)
-                  IF (tau_f) THEN
-                     CALL calc_tau_angular(grid_atom, harmonics, nspins, ir, &
-                                           trho_h, trho_s, tau_h, tau_s)
-                  END IF
                   IF (donlcc) THEN
                      CALL calc_rho_nlcc(grid_atom, nspins, gradient_f, &
                                         ir, rho_nlcc(:, 1), rho_h, rho_s, rho_nlcc(:, 2), drho_h, drho_s)
                   END IF
-                  CALL fill_rho_set(rho_set_h, lsd, nspins, needs, rho_h, drho_h, tau_h, na, ir)
-                  CALL fill_rho_set(rho_set_s, lsd, nspins, needs, rho_s, drho_s, tau_s, na, ir)
+                  CALL fill_rho_set(rho_set_h, lsd, nspins, needs, rho_h, drho_h, tau_h(:, ir, :), na, ir)
+                  CALL fill_rho_set(rho_set_s, lsd, nspins, needs, rho_s, drho_s, tau_s(:, ir, :), na, ir)
```

Comment:
this will generate temporaries, right? Just to be aware of it.

Reply:
I did not think about it when coding, but yes, I think it will generate temporaries. I don't see how to address it without changing the overall flow though.